lib/interfaces/src/ADCInterface.cpp

```diff
@@ -0,0 +1,130 @@
+  adc1_scales.at(_adc_parameters.channels.thermistor4_channel)        = _adc_parameters.scales.thermistor4_scale;
+  adc1_scales.at(_adc_parameters.channels.thermistor5_channel)        = _adc_parameters.scales.thermistor5_scale;
+  adc1_scales.at(_adc_parameters.channels.thermistor6_channel)        = _adc_parameters.scales.thermistor6_scale;
+  adc1_scales.at(_adc_parameters.channels.thermistor7_channel)        = _adc_parameters.scales.thermistor7_scale;
+
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor0_offset;
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor1_offset;
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor2_offset;
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor3_offset;
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor4_offset;
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor5_offset;
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor6_offset;
+  adc1_offsets.at(_adc_parameters.channels.thermistor0_channel)        = _adc_parameters.offsets.thermistor7_offset;
+
+  _adc0.emplace(
+    _adc_parameters.pinouts.adc0_spi_cs_pin, 
+    MCP_ADC_DEFAULT_SPI_SDI, 
+    MCP_ADC_DEFAULT_SPI_SDO, 
+    MCP_ADC_DEFAULT_SPI_CLK,
+    MCP_ADC_DEFAULT_SPI_SPEED,
+    adc0_scales.data(),
+    adc0_offsets.data()
+  );
+
+  _adc1.emplace(
+    _adc_parameters.pinouts.adc1_spi_cs_pin, 
+    MCP_ADC_DEFAULT_SPI_SDI, 
+    MCP_ADC_DEFAULT_SPI_SDO, 
+    MCP_ADC_DEFAULT_SPI_CLK,
+    MCP_ADC_DEFAULT_SPI_SPEED,
+    adc1_scales.data(),
+    adc1_offsets.data()
+  );
+}
+
+void ADCInterface::tick_adc0() { _adc0->tick(); }
+void ADCInterface::tick_adc1() { _adc1->tick(); }
+
+const ADCInterfaceParams_s& ADCInterface::get_adc_params() const {
+    return _adc_parameters;
+}
+
+AnalogConversion_s ADCInterface::read_glv() {
+  return _adc0->data.conversions.at(_adc_parameters.channels.glv_sense_channel);
+}
+
+AnalogConversion_s ADCInterface::read_bspd_current() {
+  return _adc0->data.conversions.at(_adc_parameters.channels.current_sense_channel);
+}
+
+AnalogConversion_s ADCInterface::read_bspd_reference_current() {
+  return _adc0->data.conversions.at(_adc_parameters.channels.reference_sense_channel);
+}
+
+AnalogConversion_s ADCInterface::read_rl_loadcell() {
+  return _adc0->data.conversions.at(_adc_parameters.channels.rl_loadcell_channel);
+}
+
+AnalogConversion_s ADCInterface::read_rr_loadcell() {
+  return _adc0->data.conversions.at(_adc_parameters.channels.rr_loadcell_channel);
+}
+
+AnalogConversion_s ADCInterface::read_rl_sus_pot() {
```

Comment:
also, if we're gonna use optionals that's fine but we should do the neccesary checking to make sure it's safe. we don't want to put any burden of checking on the user. 

`if (!_adc0) {do something}`

Reply:
we could probably add a accessor guard so the adc interface does all the safety checking:

```cpp
public:
  MCP_ADC& adc0() {
    if(!_adc0) { fault(); return; }
    return _adc0;  
  }
```

i think the question is what exactly should `fault()` do?